### PR TITLE
Quiz editing capability for not yet activated quizzes

### LINKED ISSUE #12150: Allow coaches to edit non-assigned quizzes.
## Observed behavior
Per the discussion in https://github.com/learningequality/kolibri/pull/12111#issuecomment-2100740029, @radinamatic noticed during testing that the previously intended "a coach cannot edit a quiz" behavior is still valid with EQM, and that this can be a frustrating experience for coaches. 

For the MVP, we will allow a basic workflow where a coach is able to edit a quiz before it is assigned (this prevents opening a can of worms around versioning). This will require some design updates to allow the workflow, as well as some updates to enable this.

## Expected behavior
If a quiz is not yet visible to learners, a coach should be able to edit the quiz.

### PR DESCRIPTION
**Pull request by rtibbles**
### Backend tech debt/cleanup:
* Adds a limit kwarg to the serialize method to allow the queryset to be sliced after annotation but before a DB fetch has occurred - this probably shouldn't affect most uses of serialize object, but is added as it is useful for changes made here.
* Updates the lesson endpoint to return (active, assignments) rather than (is_active, lesson_assignments) to be parallel to the exam endpoint
* Update the class summary API endpoint to reuse the exam and lesson endpoints to avoid subtle differences between the two (leave the additional 'groups' processing for now)

### Frontend tech debt/cleanup:
* Does some cleanup of the useQuizCreation composable to stop provide/injecting unused items.
* Removes the endlessly repeated authorization props for the CoachAppBar and CoachImmersive pages, and instead just reads the values directly from the store.
* Update the frontend to use lesson.active instead of lesson.is_active
* Update the frontend to always use lesson.assignments (regardless of whether reading from the lessons API or the class summary data)
* Update the data helpers getters to work with the simplifications to data representation for exams and lessons
* Refactor the AssignmentDetailsModal to make its appearance parallel to the new title editing for quizzes and simplify the API to just accept an assignment object (leveraging the parallel data representation for exams and lessons).
* Remove unneeded slot that further complicated the API, and hide the bottom app bar when editing quizzes, instead making it expose an `update` event to allow it to be used more like a two way data binding component
* Fixes regression caused by an unreferenced usage of the common coach strings mixin
* Updates the CoachExamPage to use composables
* Cleans up unused components, routes, and vuex state after the feature work

### Backend feature work:
* Adds a new DraftExam model to store data about Exams that are not ready to be synced yet.
* Updates the API endpoint for the Exam model to handle both models, returning data for both in fetches, and flexibly updating the models depending on whether the exam is or is not now in draft
* Adds extensive API endpoint tests to cover the different cases

### Frontend feature work:
* Update the quiz creation experience to allow it to handle editing of existing quizzes - if the quiz is a draft, it allows total editing, otherwise just the title and assignments.
* Updates CoachExamPage to undraft draft quizzes when activating them
* Updates the quiz creation workflow to allow assignment of the quizzes when they are being created, using the updated AssignmentDetailsModal component
* Update all places where a quiz is edited (e.g. the 'edit details' flow) to use the new creation flow for a consistent editing experience
* Updates the copy workflow to work with the updated backend

## References
Fixes #12150
Handles some aspects of #11615 inasmuch as it deletes all the vuex state specific to this route, and switches to composables

## Reviewer guidance
Create a quiz.
Edit the quiz before activating.
Activate the quiz.
Edit after activation.
Copy quiz.
Edit a copied quiz.

----

## Testing checklist

- [ ] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Critical and brittle code paths are covered by unit tests

## PR process

- [ ] PR has the correct target branch and milestone
- [ ] PR has 'needs review' or 'work-in-progress' label
- [ ] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

## Reviewer checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

### CODE REVIEW

File: kolibri/core/exams/models.py
Comment (nucleogenesis) [inline kolibri/core/exams/models.py:167]: Are `get_*` methods automagically made into getters for the field of the same name? So `self.questions` is the same as calling `self.get_questions()`?

File: kolibri/core/exams/serializers.py
Comment (nucleogenesis) [inline kolibri/core/exams/serializers.py:224]: Did you consider doing this in `to_exam()` if the exam saves successfully? When I read that code I was thinking maybe auto-deletion should be enforced in the model. Nothing blocking just something that came to mind.

File: kolibri/plugins/coach/class_summary_api.py
Comment (nucleogenesis) [inline kolibri/plugins/coach/class_summary_api.py:41]: Could you add a comment here explaining what is happening here? Like -- why are we instantiating them at the module level? Does this only matter for the life of a single request or are the instances used across multiple requests to some benefit?

File: kolibri/plugins/coach/assets/src/views/reports/ReportsLessonListPage.vue
Comment (nucleogenesis) [inline kolibri/plugins/coach/assets/src/views/reports/ReportsLessonListPage.vue:211]: I peeked this file and there doesn't appear to be any definitions for a `getRecipientNamesForLesson` in the file. Not sure if I'm missing something here.

Comment (nucleogenesis) [reply kolibri/plugins/coach/assets/src/views/reports/ReportsLessonListPage.vue:211]: Answer: Mixin-ed

**Comment by marcellamaki** [comment]
A few follow up items as discussed during the code review walk through (hopefully comprehensible reminders to Richard): 
- There is an edge case in regards to the created date for Exams that are created via the API endpoint that should be handled 
- Is `limit=1` necessary?
- Add code comments, particularly in the `serializers, although really throughout commit `a2503ab` would be helpful to add additional details and reasoning
- [objectWithDefaults](https://github.com/learningequality/kolibri/pull/12232/commits/d3a5d30b64b07885121a687ee1ae17ff835332eb#diff-d2d69f8258c0e50627f7bb8b2f35f97c43420419acd580827737749a395fbb23R295) check does not have to happen twice
- At some point we discussed updating to `attrs.get("active")` which seems like it would have been in a serializer but now I cannot find the line in the diff.

**Comment by rtibbles** [comment]
Hopefully I've addressed all the issues we identified @nucleogenesis @marcellamaki - I did a rebase, so as to keep the diff clean and drop the unneeded commit.

**Comment by pcenov** [comment]
Hi @rtibbles it's not possible to save a quiz - I'm getting a 500 server error when I attempt to do so:

https://github.com/learningequality/kolibri/assets/79847249/18d8f523-90f0-469f-8e4e-51c56a93933b

Logs: [Logs.zip](https://github.com/user-attachments/files/15546326/Logs.zip)

**Comment by rtibbles** [comment]
Ah yes, this is the result of my updating things based on yesterday's code review, and then only testing the editing flow, not the creation flow, afterwards!

**Review by nucleogenesis — COMMENTED** [review]
✅ Code review

I'll leave PR approval until QA has given it a pass. Once we're cleared of regressions we'll merge this.

**Comment by pcenov** [comment]
Hi @rtibbles, here are the issues I able to identify:

1. When creating/editing a quiz, clicking the Save button results in immediately closing the modal:

https://github.com/learningequality/kolibri/assets/79847249/cf2727b2-bdee-45db-9bb2-92feaa915cc6

2. When editing a quiz I cannot replace a question because clicking the checkbox of a question results in selecting all the questions:

https://github.com/learningequality/kolibri/assets/79847249/d5219dc0-0636-4126-96eb-75ef65528075

3. Cannot edit the title of a started quiz - getting a 404 error:

https://github.com/learningequality/kolibri/assets/79847249/32271fef-de07-4105-a588-1be7d01208dc

4. 'Select a quiz' is not working, getting a `TypeError: Cannot read properties of undefined (reading 'data')`:

https://github.com/learningequality/kolibri/assets/79847249/75d455f6-1cd5-451d-a405-a253f4f0d423

5. When editing a quiz the 'Question order' value is always displayed as 'Randomized'. Note that it's not clear what should happen if I have both sections with randomized and fixed order:

https://github.com/learningequality/kolibri/assets/79847249/7d130fd4-fe55-41df-8a96-aeab91553626

**Comment by rtibbles** [comment]
Hi @pcenov - thanks for this review, only 3 is directly related to the changes here, so I'll address that within the scope of this PR, but leave the others for follow up, as they are pre-existing to my work here.

**Comment by radinamatic** [comment]
I did not do such an extensive testing as @pcenov , but I can replicate 2. (_replacing a question does not work because clicking the checkbox of a question results in selecting all the questions_), which got me into a 400 error:

400-patch

**Review by nucleogenesis — APPROVED** [review]
Blocking changes have been addressed, good to merge once follow-ups are made! Thanks Richard

**Comment by nucleogenesis** [comment]
@rtibbles @radinamatic I've tested this locally and the code LGTM

**Comment by rtibbles** [comment]
@radinamatic @nucleogenesis confirmed the fix of the final issue of activating a quiz from the summary page, we're going to merge here to keep development going, and we're confident of the fix.

**Pull request merged**
